drivers/comparator/comparator_stm32_comp.c

```diff
@@ -0,0 +1,318 @@
+#define STM32_COMP_DT_INST_COMP_CONFIG_INIT(inst)				\
+	{									\
+		.InputPlus = STM32_COMP_DT_INST_P_IN(inst),			\
+		.InputMinus = STM32_COMP_DT_INST_N_IN(inst),			\
+		.InputHysteresis = STM32_COMP_DT_INST_HYST_MODE(inst),		\
+		.OutputPolarity = STM32_COMP_DT_INST_INV_OUT(inst),		\
+		.OutputBlankingSource = STM32_COMP_DT_INST_BLANK_SEL(inst),	\
+	}
+#endif /* st_stm32h7_comp */
+
+struct stm32_comp_config {
+	COMP_TypeDef *comp;
+	struct stm32_pclken *pclken;
+	const struct pinctrl_dev_config *pincfg;
+	void (*irq_init)(void);
+	const uint32_t irq_nr;
+	const LL_COMP_InitTypeDef comp_config;
+	const uint32_t exti_line_number;
+	const bool lock_enable;
+	const bool miller_effect_hold_enable;
+};
+
+struct stm32_comp_data {
+	comparator_callback_t callback;
+	void *user_data;
+};
+
+__maybe_unused static bool stm32_comp_is_resumed(void)
+{
+#if CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	(void)pm_device_state_get(DEVICE_DT_INST_GET(0), &state);
+	return state == PM_DEVICE_STATE_ACTIVE;
+#else
+	return true;
+#endif /* CONFIG_PM_DEVICE */
+}
+
+static int stm32_comp_get_output(const struct device *dev)
+{
+	const struct stm32_comp_config *cfg = dev->config;
+
+	return LL_COMP_ReadOutputLevel(cfg->comp);
+}
+
+static int stm32_comp_set_trigger(const struct device *dev, enum comparator_trigger trigger)
+{
+	const struct stm32_comp_config *cfg = dev->config;
+	struct stm32_comp_data *data = dev->data;
+	stm32_exti_trigger_type exti_trigger = 0U;
+	int ret = 0;
+
+	switch (trigger) {
+	case COMPARATOR_TRIGGER_NONE:
+		exti_trigger = STM32_EXTI_TRIG_NONE;
+		break;
+	case COMPARATOR_TRIGGER_RISING_EDGE:
+		exti_trigger = STM32_EXTI_TRIG_RISING;
+		break;
+	case COMPARATOR_TRIGGER_FALLING_EDGE:
+		exti_trigger = STM32_EXTI_TRIG_FALLING;
+		break;
+	case COMPARATOR_TRIGGER_BOTH_EDGES:
+		exti_trigger = STM32_EXTI_TRIG_BOTH;
+		break;
+	default:
+		LOG_DBG("%s: Unsupported trigger mode %d", dev->name, trigger);
+		return -ENOTSUP;
+	}
+
+	irq_disable(cfg->irq_nr);
+	LL_COMP_Disable(cfg->comp);
+
+	ret = stm32_exti_enable(cfg->exti_line_number, exti_trigger,
+				STM32_EXTI_MODE_IT);
+	if (ret != 0) {
+		LOG_DBG("%s: EXTI init failed (%d)", dev->name, ret);
+		return ret;
+	}
+
+	if (stm32_comp_is_resumed()) {
+		LL_COMP_Enable(cfg->comp);
+	}
+
+	if (data->callback != NULL) {
+		irq_enable(cfg->irq_nr);
+	}
+
+	return ret;
+}
+
+static int stm32_comp_trigger_is_pending(const struct device *dev)
+{
+	const struct stm32_comp_config *cfg = dev->config;
+
+	if (stm32_exti_is_pending(cfg->exti_line_number)) {
+		stm32_exti_clear_pending(cfg->exti_line_number);
+		return 1;
+	}
+	return 0;
+}
+
+static int stm32_comp_set_trigger_callback(const struct device *dev, comparator_callback_t callback,
+					   void *user_data)
+{
+	const struct stm32_comp_config *cfg = dev->config;
+	struct stm32_comp_data *data = dev->data;
+
+	irq_disable(cfg->irq_nr);
+
+	data->callback = callback;
+	data->user_data = user_data;
+
+	irq_enable(cfg->irq_nr);
```

Comment:
If `callback` is NULL, I think the interrupt should not be enabled.

Reply:
It makes sense at the first look. I'll take care of it.

Reply:
added callback NULL check.

Reply:
UPD: Proposed change could not be applied, since IRQ enable/disable marks only critical section here and shall be always present to assign values to the data structure. As a consequence - the callback test of COMP with proposed changes will fail. Feel free to test it your own.